Document missing color components (#27523)
mdn/content: files/en-us/web/css/color_value/index.md

{{EmbedLiveSample("currentcolor_keyword", "100%", 80)}}

### Missing color components

Each component of any CSS color functions - except for those using the legacy comma-separated syntax - can be specified as the keyword `none` to be a missing component.

> Explicitly specifying missing components is usually useful in [color interpolation](#interpolation_with_missing_components). For all other purposes, a missing component will effectively have a zero value in an appropriate unit: `0`, `0%`, or `0deg`. For example:

Reviewer: This is good, we could also add why and a small clarification that the examples are in isolation and not used in interpolation:

```suggestion
Explicitly specifying missing components is useful in [color interpolation](#interpolation_with_missing_components) for cases where you would like to interpolate some color components but not others. For all other purposes, a missing component will effectively have a zero value in an appropriate unit: `0`, `0%`, or `0deg`. For example, the following colors are equivalent when used outside of interpolation:
```

What do you think?

Author: I agree that uplifting the use case to this section is more intuitive :+1:

```css

> oklab(50% none -0.25) /* is equivalent to */
> oklab(50% 0 -0.25)
> 
> hsl(none 100% 50%) /* is equivalent to */
> hsl(0deg 100% 50%)

Reviewer: What do you think about adding the whole declarations to show we're not doing interpolation?

```suggestion
/* These are equivalent */
background-color: oklab(50% none -0.25);
background-color: oklab(50% 0 -0.25);

/* These are equivalent */
background-color: hsl(none 100% 50%);
background-color: hsl(0deg 100% 50%);
```

Author: In addition, I altered the CSS property in the first group of examples for variety.

```css
/* These are equivalent */
color: oklab(50% none -0.25);
color: oklab(50% 0 -0.25);

/* These are equivalent */
background-color: hsl(none 100% 50%);
background-color: hsl(0deg 100% 50%);
```

## Interpolation

Color interpolation happens with [gradients](/en-US/docs/Web/CSS/gradient), [transitions](/en-US/docs/Web/CSS/CSS_transitions/Using_CSS_transitions), and [animations](/en-US/docs/Web/CSS/CSS_animations/Using_CSS_animations).

When interpolating `<color>` values, they are first converted to a given color space, and then each component of the [computed values](/en-US/docs/Web/CSS/computed_value) are interpolated linearly, with interpolation's speed being determined by the [timing function](/en-US/docs/Web/CSS/easing-function) in transitions and animations. The interpolation color space defaults to Oklab, but can be overriden through {{CSSXref("&lt;color-interpolation-method&gt;")}} in some color-related functional notations.

### Interpolation with missing components

> #### Simple case

Reviewer: Editorial comment; in general we're trying to avoid using "simple" as this is subjective. We could change to something to do with `color-mix()`, like:

```suggestion
#### Mixing colors
```

> In the simple case, both colors to be interpolated are already in the interpolation color space, e.g.:
> 
> ```css
> color-mix(in oklch, oklch(none 0.2 10), oklch(60% none 30))
> ```
> 
> Before the interpolation takes place, any missing component will be replaced with the same component from the other color. Thus the expression above is equivalent to:
> 
> ```css
> color-mix(in oklch, oklch(60% 0.2 10), oklch(60% 0.2 30))
> ```

Reviewer: ````suggestion
When interpolating colors that are in the same color space, missing components are replaced with the component value from the other color.
For example, the following two expressions are equivalent:

```css
color-mix(in oklch, oklch(none 0.2 10), oklch(60% none 30))
color-mix(in oklch, oklch(60% 0.2 10), oklch(60% 0.2 30))
```
````

> **Note:** If a component is missing from both colors, this component will be missing after the interpolation.

> #### General case

Reviewer: I'm not sure if `types` adds confusion, but I think this describes what we're doing in this section:

```suggestion
#### Equivalent component types
```

Reviewer: Or alternatively, this happens when we're mixing different types of color values or functions, i.e.:

```
color-mix(hsl(none 80% 40%), lch(60% none 30))
```

So we could say "Converting between color types", what do you think?

Author: My intention for having two subsections here was to flatten the learning curve by presenting the basic yet most common case first. (The spec effectively only has the former "general case".) To remove the subjectivity while keeping this goal, the two headings have been changed with "analogous components" named. Are they OK now?

Reviewer: Yes, looks perfect now. Let's merge shortly

In the general case, missing components are first transferred into the converted colors based the analogy of components. In the following table, components of the same category are **analogous components**.

| Category     | Components |
| ------------ | ---------- |
| Reds         | `R`, `X`   |
| Greens       | `G`, `Y`   |
| Blues        | `B`, `Z`   |
| Lightness    | `L`        |
| Colorfulness | `C`, `S`   |
| Hue          | `H`        |
| a            | `a`        |
| b            | `b`        |

For example:

- `X` (`0.2`) in `color(xyz 0.2 0.1 0.6)` is analogous to `R` (`50%`) in `rgb(50% 70% 30%)`.
- `H` (`0deg`) in `hsl(0deg 100% 80%)` is analogous to `H` (`140`) in `oklch(80% 0.1 140)`.

Using Oklch as the interpolation color space and the two colors below as an example:

```css
lch(80% 30 none)
color(display-p3 0.7 0.5 none)
```

> the preprocessing procedure is:

Reviewer: This is a really good addition 👍🏻
